service/controller/resource/subnet/resource.go

```diff
@@ -0,0 +1,273 @@
+package subnet
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	azureresource "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller/context/reconciliationcanceledcontext"
+	corev1 "k8s.io/api/core/v1"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/client"
+	"github.com/giantswarm/azure-operator/v4/pkg/label"
+	"github.com/giantswarm/azure-operator/v4/service/controller/debugger"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+	subnet "github.com/giantswarm/azure-operator/v4/service/controller/resource/subnet/template"
+)
+
+const (
+	credentialDefaultName = "credential-default"
+	credentialNamespace   = "giantswarm"
+	mainDeploymentName    = "subnet"
+	// Name is the identifier of the resource.
+	Name = "subnet"
+)
+
+type Config struct {
+	AzureClientsFactory *client.Factory
+	CtrlClient          ctrlclient.Client
+	Debugger            *debugger.Debugger
+	Logger              micrologger.Logger
+}
+
+type Resource struct {
+	azureClientsFactory *client.Factory
+	ctrlClient          ctrlclient.Client
+	debugger            *debugger.Debugger
+	logger              micrologger.Logger
+}
+
+type StorageAccountIpRule struct {
+	Value  string `json:"value"`
+	Action string `json:"action"`
+}
+
+func New(config Config) (*Resource, error) {
+	if config.AzureClientsFactory == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AzureClientsFactory must not be empty", config)
+	}
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Debugger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Debugger must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		azureClientsFactory: config.AzureClientsFactory,
+		ctrlClient:          config.CtrlClient,
+		debugger:            config.Debugger,
+		logger:              config.Logger,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToAzureCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, &cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deploymentsClient, err := r.azureClientsFactory.GetDeploymentsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	armTemplate, err := subnet.GetARMTemplate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, allocatedSubnet := range cr.Spec.NetworkSpec.Subnets {
```

Comment:
For every subnet declared in the `AzureCluster.Spec.NetworkSpec.Subnets` field, we submit a deployment to Azure to create the subnet.

Reply:
> This handler makes no assumptions, it just creates the subnet using all the details from AzureCluster.

Love this! :rocket: